File: pkg/operations/adb.go
package operations

import (
	"context"
	"fmt"
	"os/exec"
	"strconv"
	"strings"
	"time"
)

// ADBHandler handles ADB operations
type ADBHandler struct {
	deviceID string
}

// NewADBHandler creates a new ADB handler
func NewADBHandler() *ADBHandler {
	return &ADBHandler{}
}

// SetDevice sets the target device ID
func (h *ADBHandler) SetDevice(deviceID string) {
	h.deviceID = deviceID
}

// GetType returns the operation type
func (h *ADBHandler) GetType() string {
	return "adb"
}

// Validate checks if the parameters are valid
func (h *ADBHandler) Validate(params map[string]interface{}) error {
	action, ok := params["action"].(string)
	if !ok {
		return fmt.Errorf("action parameter is required")
	}

	switch action {
	case "tap":
		if _, ok := params["x"]; !ok {
			return fmt.Errorf("x parameter is required for tap action")
		}
		if _, ok := params["y"]; !ok {
			return fmt.Errorf("y parameter is required for tap action")
		}
	case "swipe":
		if _, ok := params["x1"]; !ok {
			return fmt.Errorf("x1 parameter is required for swipe action")
		}
		if _, ok := params["y1"]; !ok {
			return fmt.Errorf("y1 parameter is required for swipe action")
		}
		if _, ok := params["x2"]; !ok {
			return fmt.Errorf("x2 parameter is required for swipe action")
		}
		if _, ok := params["y2"]; !ok {
			return fmt.Errorf("y2 parameter is required for swipe action")
		}
	case "input":
		if _, ok := params["text"]; !ok {
			return fmt.Errorf("text parameter is required for input action")
		}
	case "install":
		if _, ok := params["apk_path"]; !ok {
			return fmt.Errorf("apk_path parameter is required for install action")
		}
	case "uninstall":
		if _, ok := params["package"]; !ok {
			return fmt.Errorf("package parameter is required for uninstall action")
		}
	case "screencap":
		if _, ok := params["output_path"]; !ok {
			return fmt.Errorf("output_path parameter is required for screencap action")
		}
	case "shell":
		if _, ok := params["command"]; !ok {
			return fmt.Errorf("command parameter is required for shell action")
		}
	}

	return nil
}

// Execute performs the ADB operation
func (h *ADBHandler) Execute(ctx context.Context, params map[string]interface{}, execCtx interface{}) (map[string]interface{}, error) {
	action := params["action"].(string)

	// Build base command
	args := []string{}
	if h.deviceID != "" {
		args = append(args, "-s", h.deviceID)
	}

	var result map[string]interface{}
	var err error

	switch action {
	case "tap":
		result, err = h.executeTap(ctx, args, params)
	case "swipe":
		result, err = h.executeSwipe(ctx, args, params)
	case "input":
		result, err = h.executeInput(ctx, args, params)
	case "install":
		result, err = h.executeInstall(ctx, args, params)
	case "uninstall":
		result, err = h.executeUninstall(ctx, args, params)
	case "screencap":
		result, err = h.executeScreencap(ctx, args, params)
	case "shell":
		result, err = h.executeShell(ctx, args, params)
	case "click":
		result, err = h.executeClick(ctx, args, params)
	case "press":
		result, err = h.executePress(ctx, args, params)
	default:
		return nil, fmt.Errorf("unsupported ADB action: %s", action)
	}

	return result, err
}

func (h *ADBHandler) executeTap(ctx context.Context, baseArgs []string, params map[string]interface{}) (map[string]interface{}, error) {
	x := getIntParam(params, "x")
	y := getIntParam(params, "y")

	args := append(baseArgs, "shell", "input", "tap", strconv.Itoa(x), strconv.Itoa(y))
	cmd := exec.CommandContext(ctx, "adb", args...)

	output, err := cmd.CombinedOutput()
	if err != nil {
		return nil, fmt.Errorf("tap failed: %v, output: %s", err, string(output))
	}

	return map[string]interface{}{
		"success": true,
		"x":       x,
		"y":       y,
	}, nil
}

func (h *ADBHandler) executeSwipe(ctx context.Context, baseArgs []string, params map[string]interface{}) (map[string]interface{}, error) {
	x1 := getIntParam(params, "x1")
	y1 := getIntParam(params, "y1")
	x2 := getIntParam(params, "x2")
	y2 := getIntParam(params, "y2")
	duration := getIntParamDefault(params, "duration", 300)

	args := append(baseArgs, "shell", "input", "swipe",
		strconv.Itoa(x1), strconv.Itoa(y1),
		strconv.Itoa(x2), strconv.Itoa(y2),
		strconv.Itoa(duration))
	cmd := exec.CommandContext(ctx, "adb", args...)

	output, err := cmd.CombinedOutput()
	if err != nil {
		return nil, fmt.Errorf("swipe failed: %v, output: %s", err, string(output))
	}

	return map[string]interface{}{
		"success":  true,
		"from":     map[string]int{"x": x1, "y": y1},
		"to":       map[string]int{"x": x2, "y": y2},
		"duration": duration,
	}, nil
}

func (h *ADBHandler) executeInput(ctx context.Context, baseArgs []string, params map[string]interface{}) (map[string]interface{}, error) {
	text := getStringParam(params, "text")

	args := append(baseArgs, "shell", "input", "text", text)
	cmd := exec.CommandContext(ctx, "adb", args...)

	output, err := cmd.CombinedOutput()
	if err != nil {
		return nil, fmt.Errorf("input failed: %v, output: %s", err, string(output))
	}

	return map[string]interface{}{
		"success": true,
		"text":    text,
	}, nil
}

func (h *ADBHandler) executeInstall(ctx context.Context, baseArgs []string, params map[string]interface{}) (map[string]interface{}, error) {
	apkPath := getStringParam(params, "apk_path")

	args := append(baseArgs, "install", apkPath)
	cmd := exec.CommandContext(ctx, "adb", args...)

	output, err := cmd.CombinedOutput()
	if err != nil {
		return nil, fmt.Errorf("install failed: %v, output: %s", err, string(output))
	}

	return map[string]interface{}{
		"success": true,
		"apk":     apkPath,
		"output":  string(output),
	}, nil
}

func (h *ADBHandler) executeUninstall(ctx context.Context, baseArgs []string, params map[string]interface{}) (map[string]interface{}, error) {
	pkg := getStringParam(params, "package")

	args := append(baseArgs, "uninstall", pkg)
	cmd := exec.CommandContext(ctx, "adb", args...)

	output, err := cmd.CombinedOutput()
	if err != nil {
		return nil, fmt.Errorf("uninstall failed: %v, output: %s", err, string(output))
	}

	return map[string]interface{}{
		"success": true,
		"package": pkg,
		"output":  string(output),
	}, nil
}

func (h *ADBHandler) executeScreencap(ctx context.Context, baseArgs []string, params map[string]interface{}) (map[string]interface{}, error) {
	outputPath := getStringParam(params, "output_path")

	args := append(baseArgs, "shell", "screencap", "-p", "/sdcard/screenshot.png")
	cmd := exec.CommandContext(ctx, "adb", args...)

	output, err := cmd.CombinedOutput()
	if err != nil {
		return nil, fmt.Errorf("screencap failed: %v, output: %s", err, string(output))
	}

	// Pull the screenshot
	pullArgs := append(baseArgs, "pull", "/sdcard/screenshot.png", outputPath)
	pullCmd := exec.CommandContext(ctx, "adb", pullArgs...)

	pullOutput, err := pullCmd.CombinedOutput()
	if err != nil {
		return nil, fmt.Errorf("pull screenshot failed: %v, output: %s", err, string(pullOutput))
	}

	return map[string]interface{}{
		"success":     true,
		"output_path": outputPath,
	}, nil
}

func (h *ADBHandler) executeShell(ctx context.Context, baseArgs []string, params map[string]interface{}) (map[string]interface{}, error) {
	command := getStringParam(params, "command")

	args := append(baseArgs, "shell", command)
	cmd := exec.CommandContext(ctx, "adb", args...)

	output, err := cmd.CombinedOutput()
	if err != nil {
		return nil, fmt.Errorf("shell command failed: %v, output: %s", err, string(output))
	}

	return map[string]interface{}{
		"success": true,
		"command": command,
		"output":  string(output),
	}, nil
}

func (h *ADBHandler) executeClick(ctx context.Context, baseArgs []string, params map[string]interface{}) (map[string]interface{}, error) {
	x := getIntParam(params, "x")
	y := getIntParam(params, "y")

	args := append(baseArgs, "shell", "input", "tap", strconv.Itoa(x), strconv.Itoa(y))
	cmd := exec.CommandContext(ctx, "adb", args...)

	output, err := cmd.CombinedOutput()
	if err != nil {
		return nil, fmt.Errorf("click failed: %v, output: %s", err, string(output))
	}

	return map[string]interface{}{
		"success": true,
		"x":       x,
		"y":       y,
	}, nil
}

func (h *ADBHandler) executePress(ctx context.Context, baseArgs []string, params map[string]interface{}) (map[string]interface{}, error) {
	key := getStringParam(params, "key")

	args := append(baseArgs, "shell", "input", "keyevent", key)
	cmd := exec.CommandContext(ctx, "adb", args...)

	output, err := cmd.CombinedOutput()
	if err != nil {
		return nil, fmt.Errorf("press failed: %v, output: %s", err, string(output))
	}

	return map[string]interface{}{
		"success": true,
		"key":     key,
	}, nil
}

// Helper functions
func getIntParam(params map[string]interface{}, key string) int {
	if val, ok := params[key]; ok {
		switch v := val.(type) {
		case float64:
			return int(v)
		case int:
			return v
		case string:
			if i, err := strconv.Atoi(v); err == nil {
				return i
			}
		}
	}
	return 0
}

func getIntParamDefault(params map[string]interface{}, key string, defaultValue int) int {
	if val, ok := params[key]; ok {
		switch v := val.(type) {
		case float64:
			return int(v)
		case int:
			return v
		case string:
			if i, err := strconv.Atoi(v); err == nil {
				return i
			}
		}
	}
	return defaultValue
}

func getStringParam(params map[string]interface{}, key string) string {
	if val, ok := params[key]; ok {
		if str, ok := val.(string); ok {
			return str
		}
	}
	return ""
}

// WaitForDevice waits for device to be connected
func (h *ADBHandler) WaitForDevice(timeout time.Duration) error {
	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()

	args := []string{}
	if h.deviceID != "" {
		args = append(args, "-s", h.deviceID)
	}
	args = append(args, "wait-for-device")

	cmd := exec.CommandContext(ctx, "adb", args...)
	return cmd.Run()
}

// GetConnectedDevices returns list of connected devices
func (h *ADBHandler) GetConnectedDevices() ([]string, error) {
	cmd := exec.Command("adb", "devices")
	output, err := cmd.CombinedOutput()
	if err != nil {
		return nil, fmt.Errorf("failed to get devices: %v", err)
	}

	lines := strings.Split(string(output), "\n")
	devices := []string{}

	for _, line := range lines[1:] {
		line = strings.TrimSpace(line)
		if line != "" && !strings.Contains(line, "daemon") {
			parts := strings.Fields(line)
			if len(parts) >= 2 && parts[1] == "device" {
				devices = append(devices, parts[0])
			}
		}
	}

	return devices, nil
}

File: src/operations/adb.py
"""
ADB operations handler for the auto-workflow system.
"""
import subprocess
import time
from typing import Dict, Any, List
from .handler import Handler
from ..models import ExecutionContext


class ADBHandler(Handler):
    """Handles ADB operations."""

    def __init__(self):
        self.device_id: str = ""

    def set_device(self, device_id: str) -> None:
        """Set the target device ID."""
        self.device_id = device_id

    def get_type(self) -> str:
        """Return the operation type."""
        return "adb"

    def validate(self, params: Dict[str, Any]) -> None:
        """Check if the parameters are valid."""
        action = params.get("action")
        if not action or not isinstance(action, str):
            raise ValueError("action parameter is required")

        if action == "tap":
            if "x" not in params:
                raise ValueError("x parameter is required for tap action")
            if "y" not in params:
                raise ValueError("y parameter is required for tap action")
        elif action == "swipe":
            if "x1" not in params:
                raise ValueError("x1 parameter is required for swipe action")
            if "y1" not in params:
                raise ValueError("y1 parameter is required for swipe action")
            if "x2" not in params:
                raise ValueError("x2 parameter is required for swipe action")
            if "y2" not in params:
                raise ValueError("y2 parameter is required for swipe action")
        elif action == "input":
            if "text" not in params:
                raise ValueError("text parameter is required for input action")
        elif action == "install":
            if "apk_path" not in params:
                raise ValueError("apk_path parameter is required for install action")
        elif action == "uninstall":
            if "package" not in params:
                raise ValueError("package parameter is required for uninstall action")
        elif action == "screencap":
            if "output_path" not in params:
                raise ValueError("output_path parameter is required for screencap action")
        elif action == "shell":
            if "command" not in params:
                raise ValueError("command parameter is required for shell action")

    def execute(self, params: Dict[str, Any], exec_ctx: ExecutionContext) -> Dict[str, Any]:
        """Perform the ADB operation."""
        action = params["action"]

        # Build base command
        args = []
        if self.device_id:
            args.extend(["-s", self.device_id])

        if action == "tap":
            return self._execute_tap(args, params)
        elif action == "swipe":
            return self._execute_swipe(args, params)
        elif action == "input":
            return self._execute_input(args, params)
        elif action == "install":
            return self._execute_install(args, params)
        elif action == "uninstall":
            return self._execute_uninstall(args, params)
        elif action == "screencap":
            return self._execute_screencap(args, params)
        elif action == "shell":
            return self._execute_shell(args, params)
        elif action == "click":
            return self._execute_click(args, params)
        elif action == "press":
            return self._execute_press(args, params)
        else:
            raise ValueError(f"unsupported ADB action: {action}")

    def _execute_tap(self, base_args: List[str], params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute tap operation."""
        x = self._get_int_param(params, "x")
        y = self._get_int_param(params, "y")

        args = base_args + ["shell", "input", "tap", str(x), str(y)]
        result = subprocess.run(["adb"] + args, capture_output=True, text=True)

        if result.returncode != 0:
            raise RuntimeError(f"tap failed: {result.stderr}")

        return {
            "success": True,
            "x": x,
            "y": y,
        }

    def _execute_swipe(self, base_args: List[str], params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute swipe operation."""
        x1 = self._get_int_param(params, "x1")
        y1 = self._get_int_param(params, "y1")
        x2 = self._get_int_param(params, "x2")
        y2 = self._get_int_param(params, "y2")
        duration = self._get_int_param(params, "duration", 300)

        args = base_args + [
            "shell", "input", "swipe",
            str(x1), str(y1), str(x2), str(y2), str(duration)
        ]
        result = subprocess.run(["adb"] + args, capture_output=True, text=True)

        if result.returncode != 0:
            raise RuntimeError(f"swipe failed: {result.stderr}")

        return {
            "success": True,
            "from": {"x": x1, "y": y1},
            "to": {"x": x2, "y": y2},
            "duration": duration,
        }

    def _execute_input(self, base_args: List[str], params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute input operation."""
        text = self._get_string_param(params, "text")

        args = base_args + ["shell", "input", "text", text]
        result = subprocess.run(["adb"] + args, capture_output=True, text=True)

        if result.returncode != 0:
            raise RuntimeError(f"input failed: {result.stderr}")

        return {
            "success": True,
            "text": text,
        }

    def _execute_install(self, base_args: List[str], params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute install operation."""
        apk_path = self._get_string_param(params, "apk_path")

        args = base_args + ["install", apk_path]
        result = subprocess.run(["adb"] + args, capture_output=True, text=True)

        if result.returncode != 0:
            raise RuntimeError(f"install failed: {result.stderr}")

        return {
            "success": True,
            "apk": apk_path,
            "output": result.stdout,
        }

    def _execute_uninstall(self, base_args: List[str], params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute uninstall operation."""
        package = self._get_string_param(params, "package")

        args = base_args + ["uninstall", package]
        result = subprocess.run(["adb"] + args, capture_output=True, text=True)

        if result.returncode != 0:
            raise RuntimeError(f"uninstall failed: {result.stderr}")

        return {
            "success": True,
            "package": package,
            "output": result.stdout,
        }

    def _execute_screencap(self, base_args: List[str], params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute screencap operation."""
        output_path = self._get_string_param(params, "output_path")

        # Take screenshot on device
        args = base_args + ["shell", "screencap", "-p", "/sdcard/screenshot.png"]
        result = subprocess.run(["adb"] + args, capture_output=True, text=True)

        if result.returncode != 0:
            raise RuntimeError(f"screencap failed: {result.stderr}")

        # Pull screenshot to local
        pull_args = base_args + ["pull", "/sdcard/screenshot.png", output_path]
        pull_result = subprocess.run(["adb"] + pull_args, capture_output=True, text=True)

        if pull_result.returncode != 0:
            raise RuntimeError(f"pull screenshot failed: {pull_result.stderr}")

        return {
            "success": True,
            "output_path": output_path,
        }

    def _execute_shell(self, base_args: List[str], params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute shell command operation."""
        command = self._get_string_param(params, "command")

        args = base_args + ["shell", command]
        result = subprocess.run(["adb"] + args, capture_output=True, text=True)

        if result.returncode != 0:
            raise RuntimeError(f"shell command failed: {result.stderr}")

        return {
            "success": True,
            "command": command,
            "output": result.stdout,
        }

    def _execute_click(self, base_args: List[str], params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute click operation (alias for tap)."""
        x = self._get_int_param(params, "x")
        y = self._get_int_param(params, "y")

        args = base_args + ["shell", "input", "tap", str(x), str(y)]
        result = subprocess.run(["adb"] + args, capture_output=True, text=True)

        if result.returncode != 0:
            raise RuntimeError(f"click failed: {result.stderr}")

        return {
            "success": True,
            "x": x,
            "y": y,
        }

    def _execute_press(self, base_args: List[str], params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute key press operation."""
        key = self._get_string_param(params, "key")

        args = base_args + ["shell", "input", "keyevent", key]
        result = subprocess.run(["adb"] + args, capture_output=True, text=True)

        if result.returncode != 0:
            raise RuntimeError(f"press failed: {result.stderr}")

        return {
            "success": True,
            "key": key,
        }

    def wait_for_device(self, timeout: int = 30) -> None:
        """Wait for device to be connected."""
        args = []
        if self.device_id:
            args.extend(["-s", self.device_id])
        args.append("wait-for-device")

        result = subprocess.run(
            ["adb"] + args,
            capture_output=True,
            text=True,
            timeout=timeout
        )

        if result.returncode != 0:
            raise RuntimeError(f"wait for device failed: {result.stderr}")

    def get_connected_devices(self) -> List[str]:
        """Return list of connected devices."""
        result = subprocess.run(
            ["adb", "devices"],
            capture_output=True,
            text=True
        )

        if result.returncode != 0:
            raise RuntimeError(f"failed to get devices: {result.stderr}")

        devices = []
        lines = result.stdout.split("\n")

        for line in lines[1:]:
            line = line.strip()
            if line and "daemon" not in line:
                parts = line.split()
                if len(parts) >= 2 and parts[1] == "device":
                    devices.append(parts[0])

        return devices

    # Helper methods
    def _get_int_param(self, params: Dict[str, Any], key: str, default: int = 0) -> int:
        """Get integer parameter."""
        value = params.get(key)
        if value is None:
            return default
        if isinstance(value, int):
            return value
        if isinstance(value, float):
            return int(value)
        if isinstance(value, str):
            try:
                return int(value)
            except ValueError:
                return default
        return default

    def _get_string_param(self, params: Dict[str, Any], key: str) -> str:
        """Get string parameter."""
        value = params.get(key)
        if isinstance(value, str):
            return value
        return str(value) if value is not None else ""
